### ISSUE TITLE
docstore: handling provider query limitations

### ISSUE THREAD

**Issue by jba** [google/go-cloud#1769]
We need to strike a balance between limiting queries so they can run in a single RPC on all providers, and giving users a reasonable amount of expressive power.

Here are limitations we already have. I don't think any of these are problematic:
- Limited to a single collection (I believe all our providers have this limitation.)
- No OR queries. (Easy enough to run two queries.)
- No "not equal" operator. (Can usually be done as an OR query with "<" and ">".)

Some examples of other query limitations we've run into as we've dug deeper:

[DynamoDB](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html): The collection must have both a partition key and a sort key. The query requires an equality constraint on the partition key. 

It's up to Dynamo users to set up their collection with two keys; that's not really our responsibility. But it's hard to express the second constraint in a portable way. Certainly, if you can't select all documents where aNonKeyField > 5, we don't have much of a portable API. So I think we have to push this constraint into the driver constructor options somehow. Maybe we require a default partition key that we add to every query if one isn't already specified. I'm not sure.

[Firestore](https://cloud.google.com/firestore/docs/query-data/queries): If you combine an equals filter with a range filter (greater/less), you need to make an index. And you can never have two range filters on different fields.

Again, the first of these isn't our problem, but the second is serious. Would it be bad to say that you can't write a query where X > 5 and B < 3? That seems quite limiting.

I think the Firestore driver will have to support this by applying the additional filters on the client.

@shantuo @vangent

**Comment by shantuo** [comment]
Dynamo's key condition roughly equals to firestore's filter, while its filter condition works on the results returned by the query request, which means it locally filters out the result instead of send these along with the query request. But the partition key only allows "=" comparison.

FYI, I just found AWS [Document Database](https://aws.amazon.com/nosql/document/), which works like Mongo...

**Comment by eliben** [comment]
> FYI, I just found AWS [Document Database](https://aws.amazon.com/nosql/document/), which works like Mongo...

It would be nice to support it out-of-the box once our Mongo driver is working :)

**Comment by shantuo** [comment]
I think one thing we could do is to provide better error message in the dynamo driver when there is a validation expression returned.

The user can query on different fields by creating a secondary index besides their primary one. So they will need to manage their index to fit their needs. They only thing I'm afraid of is that the portable API might make these details harder to find.

**Comment by jba** [comment]
> > FYI, I just found AWS [Document Database](https://aws.amazon.com/nosql/document/), which works like Mongo...
> 
> It would be nice to support it out-of-the box once our Mongo driver is working :)

Yes, we'll get Amazon Document Database and Microsoft Cosmos for free with our Mongo driver.

**Comment by jba** [comment]
Here's another severe limitation: DynamoDB can only order by sort key (ascending or descending). That one is pretty severe, since to overcome it the client would have to store the entire result in memory.

**Comment by jba** [comment]
Oh, it gets worse: in Firestore, if you have an inequality filter (e.g. `n > 3`), then the primary sort order _must_ be on that filter's field (`n` in this case). That means we can't even specify that all queries return their results in key order.

I think we have to say that query order is unspecified.

**Comment by vangent** [comment]
> balance between limiting queries so they can run in a single
> RPC on all providers, and giving users a reasonable amount of
> expressive power

IMHO latter is better. We can document which kinds of queries may have performance issues due to client-side work per driver. IIUC DynamoDB does this already (client-side filtering).

We could also consider adding an `Option` for drivers that may have to do a lot of client-side work to return an error instead of doing so, if users want to get notified of such things early rather than via slow performance.

> Maybe we require a default partition key that we add to every
> query if one isn't already specified

+1. IIUC, we are currently generating a unique partition key (?), so this would be allowing the user to provide a value instead.

> applying the additional filters on the client

+1. We just need to document performance considerations clearly per driver.

>  think we have to say that query order is unspecified

+1. We should probably shuffle results before returning them ala `map`, to ensure users aren't counting on ordering that's driver-specific.

**Comment by jba** [comment]
After team discussion, we tentatively agreed that we should say nothing about ordering and provide no way to change it. Documented in https://github.com/google/go-cloud/pull/1774.

I also documented that we will do client-side filtering in Where clauses when necessary.

**Comment by Inuart** [comment]
`No OR queries. (Easy enough to run two queries.)`

I must say that in our case that's not so easy because our data is constantly changing. When we `Query().Where("status", "=", "A")` and then `Query().Where("status", "=", "B")`, some documents appear on both results and we need to be very careful when merging them.

**Comment by shantuo** [comment]
We cannot support that because most services don't. So we would have the same problem if we implement the `OR` queries in our drivers.

I'm not sure what operation you need with these type of queries on the constantly changing data, but it sounds like it needs some custom design/logic around the operation to get reliable results of those queries anyways.

**Comment by Inuart** [comment]
To add some context, we want a "snapshot" of how the data was at a specific moment. We don't care if we fetch slightly outdated data but we care that we have no duplicate or missing data.

**Comment by jba** [comment]
You can just take the union of the two queries. Something like:
```
docs := map[string]Document{}
for /* each document d in first query */ {
  docs[d.ID] = d
}
for /* each document d in second query */ {
   docs[d.ID] = d
}
```

**Comment by vangent** [comment]
Separately -- which service are you using as a backend? It is worth investigating whether that service supports the functionality you want.

-- If yes, then you can always use [`As`](https://gocloud.dev/concepts/as/) to get access to the underlying service-specific type and issue the native call ([example](https://godoc.org/gocloud.dev/docstore#Collection.As)). If you do this, we'd appreciate it if you'd file a separate issue with the details, because it helps us understand where the current portable API isn't covering user needs.

-- If no, then it seems like the best you can do is what @jba suggested above, possibly issuing the two queries in parallel for speed.

**Comment by Inuart** [comment]
@jba with the union I avoid getting duplicate data, but I still can "get" missing data.

If we query for documents with A and then with B instead of both at the same time we will have duplicates that go from A to B between queries and hidden ones that go from B to A between queries.

Ok @vangent, I'll file a separate issue.

**Comment by vangent** [comment]
I don't think there's much left here, see the new issue.
